src/Nonlinear/parse.jl

```diff
@@ -0,0 +1,311 @@
+    parent_var = length(expr.nodes)
+    for i in 2:length(x.args)
+        parse_expression(data, expr, x.args[i], parent_var)
+    end
+    return
+end
+
+# This function parses double inequalities like `a <= b <= c`. It's not to be
+# confused with `_parse_inequality_expression`, which handles things like
+# `a <= b`.
+function _parse_comparison_expression(
+    data::Model,
+    expr::Expression,
+    x::Expr,
+    parent_index::Int,
+)
+    for k in 2:2:length(x.args)-1
+        @assert x.args[k] == x.args[2] # don't handle a <= b >= c
+    end
+    operator_id = data.operators.comparison_operator_to_id[x.args[2]]
+    push!(expr.nodes, Node(NODE_COMPARISON, operator_id, parent_index))
+    parent_var = length(expr.nodes)
+    for i in 1:2:length(x.args)
+        parse_expression(data, expr, x.args[i], parent_var)
+    end
+    return
+end
+
+function _parse_logic_expression(
+    data::Model,
+    expr::Expression,
+    x::Expr,
+    parent_index::Int,
+)
+    id = data.operators.logic_operator_to_id[x.head]
+    push!(expr.nodes, Node(NODE_LOGIC, id, parent_index))
+    parent_var = length(expr.nodes)
+    parse_expression(data, expr, x.args[1], parent_var)
+    parse_expression(data, expr, x.args[2], parent_var)
+    return
+end
+
+function parse_expression(
+    ::Model,
+    expr::Expression,
+    x::MOI.VariableIndex,
+    parent_index::Int,
+)
+    push!(expr.nodes, Node(NODE_MOI_VARIABLE, x.value, parent_index))
+    return
+end
+
+function parse_expression(::Model, expr::Expression, x::Real, parent_index::Int)
+    push!(expr.values, convert(Float64, x)::Float64)
+    push!(expr.nodes, Node(NODE_VALUE, length(expr.values), parent_index))
+    return
+end
+
+function parse_expression(
+    ::Model,
+    expr::Expression,
+    x::ParameterIndex,
+    parent_index::Int,
+)
+    push!(expr.nodes, Node(NODE_PARAMETER, x.value, parent_index))
+    return
+end
+
+function parse_expression(
+    ::Model,
+    expr::Expression,
+    x::ExpressionIndex,
+    parent_index::Int,
+)
+    push!(expr.nodes, Node(NODE_SUBEXPRESSION, x.value, parent_index))
+    return
+end
+
+function parse_expression(
+    ::Model,
+    expr::Expression,
+    x::AbstractArray,
+    parent_index::Int,
+)
+    return error(
+        "Unexpected array $x in nonlinear expression. Nonlinear expressions " *
+        "may contain only scalar expressions.",
+    )
+end
+
+function _normalize_constraint_expr(lhs::Real, body, rhs::Real)
+    return Float64(lhs), body, Float64(rhs)
+end
+
+function _normalize_constraint_expr(lhs, body, rhs)
+    return error(
+        "Interval constraint contains non-constant left- or right-hand " *
+        "sides. Reformulate as two separate constraints, or move all " *
+        "variables into the central term.",
+    )
+end
+
+_normalize_constraint_expr(lhs, rhs::Real) = lhs, Float64(rhs)
+
+_normalize_constraint_expr(lhs, rhs) = Expr(:call, :-, lhs, rhs), 0.0
+
+function _expr_to_constraint(expr::Expr)
```

Comment:
This looks like what we have in JuMP's macro code so it would seem better to have it in JuMP so that we may be able to refactor it in the future by merging these similar code paths. What's the reason of having this in MOI ? It would be more consistent to have the user do `Nonlinear.add_constraint(model, :(sin(x) + 1), MOI.Interval(0.0, 1.0))`.
I feel it

Reply:
> What's the reason of having this in MOI ?

It was moved out of JuMP's macro code to here: https://github.com/jump-dev/JuMP.jl/pull/2955. 

So far, I have only made minimal changes to the code. So we've kept things like this. Once we merge this PR, I'll open another PR to change it, and we can decide if it's worth it.